File: FBDeviceControlTests/Tests/Unit/FBInstrumentsClientTests.m
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#import <XCTest/XCTest.h>

#import <FBControlCore/FBControlCore.h>
#import <FBDeviceControl/FBDeviceControl.h>

// Constants mirroring private definitions in FBInstrumentsClient.m
static const uint64_t TestArgumentMagic = 0x1F0;
static const uint32_t TestEmptyDictionaryKey = 10;
static const uint32_t TestObjectArgumentType = 2;
static const uint32_t TestInt32ArgumentType = 3;

#pragma mark - Private Method Exposure

@interface FBInstrumentsClient (Testing)

+ (NSData *)argumentDataForArgument:(id)argument;
+ (NSData *)argumentDataForInt32:(int32_t)value;
+ (NSData *)auxillaryDataFromArgumentsData:(nullable NSArray<NSData *> *)arguments;
+ (NSArray<id> *)objectArgumentsFromAuxillaryData:(NSData *)data error:(NSError **)error;
+ (NSData *)advanceData:(NSData *)data buffer:(void *)buffer length:(size_t)length;
+ (NSData *)advanceData:(NSData *)data dataOut:(NSData **)dataOut length:(size_t)length;

@end

#pragma mark - Test Class

@interface FBInstrumentsClientTests : XCTestCase
@end

@implementation FBInstrumentsClientTests

#pragma mark - argumentDataForArgument: Tests

- (void)testArgumentDataForArgument_WithString_ProducesValidSerializedData
{
  NSString *argument = @"com.example.testapp";

  NSData *result = [FBInstrumentsClient argumentDataForArgument:argument];

  XCTAssertNotNil(result, @"Result should not be nil for a valid string argument");
  XCTAssertGreaterThan(result.length, (NSUInteger)12, @"Result should be larger than the 12-byte header");

  uint32_t dictionaryKey = 0;
  uint32_t argumentType = 0;
  uint32_t argumentSize = 0;
  [result getBytes:&dictionaryKey range:NSMakeRange(0, sizeof(uint32_t))];
  [result getBytes:&argumentType range:NSMakeRange(4, sizeof(uint32_t))];
  [result getBytes:&argumentSize range:NSMakeRange(8, sizeof(uint32_t))];

  XCTAssertEqual(dictionaryKey, TestEmptyDictionaryKey, @"Dictionary key should match expected constant");
  XCTAssertEqual(argumentType, TestObjectArgumentType, @"Argument type should be ObjectArgumentType (2)");
  XCTAssertEqual((NSUInteger)argumentSize, result.length - 12, @"Argument size should match remaining data length");

  NSData *archivedData = [result subdataWithRange:NSMakeRange(12, argumentSize)];
  NSError *error = nil;
  NSString *decoded = [NSKeyedUnarchiver unarchivedObjectOfClass:[NSString class] fromData:archivedData error:&error];
  XCTAssertNil(error, @"Unarchiving should not produce an error");
  XCTAssertEqualObjects(decoded, argument, @"Decoded value should match the original argument");
}

- (void)testArgumentDataForArgument_WithDictionary_ProducesValidSerializedData
{
  NSDictionary *argument = @{@"StartSuspendedKey" : @YES, @"KillExisting" : @NO};

  NSData *result = [FBInstrumentsClient argumentDataForArgument:argument];

  XCTAssertNotNil(result);
  XCTAssertGreaterThan(result.length, (NSUInteger)12);

  uint32_t argumentSize = 0;
  [result getBytes:&argumentSize range:NSMakeRange(8, sizeof(uint32_t))];

  NSData *archivedData = [result subdataWithRange:NSMakeRange(12, argumentSize)];
  NSError *error = nil;
  NSDictionary *decoded = [NSKeyedUnarchiver unarchivedObjectOfClasses:[NSSet setWithArray:@[NSDictionary.class, NSNumber.class, NSString.class]] fromData:archivedData error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(decoded, argument, @"Decoded dictionary should match the original");
}

- (void)testArgumentDataForArgument_WithEmptyString_ProducesValidSerializedData
{
  NSString *argument = @"";

  NSData *result = [FBInstrumentsClient argumentDataForArgument:argument];

  XCTAssertNotNil(result);
  XCTAssertGreaterThan(result.length, (NSUInteger)12);

  uint32_t argumentSize = 0;
  [result getBytes:&argumentSize range:NSMakeRange(8, sizeof(uint32_t))];

  NSData *archivedData = [result subdataWithRange:NSMakeRange(12, argumentSize)];
  NSError *error = nil;
  NSString *decoded = [NSKeyedUnarchiver unarchivedObjectOfClass:[NSString class] fromData:archivedData error:&error];
  XCTAssertNil(error);
  XCTAssertEqualObjects(decoded, @"", @"Decoded string should be empty");
}

#pragma mark - argumentDataForInt32: Tests

- (void)testArgumentDataForInt32_WithPositiveValue_ProducesCorrectFormat
{
  int32_t value = 42;

  NSData *result = [FBInstrumentsClient argumentDataForInt32:value];

  XCTAssertEqual(result.length, (NSUInteger)12, @"Int32 argument data should be exactly 12 bytes");

  uint32_t dictionaryKey = 0;
  uint32_t argumentType = 0;
  int32_t decodedValue = 0;
  [result getBytes:&dictionaryKey range:NSMakeRange(0, sizeof(uint32_t))];
  [result getBytes:&argumentType range:NSMakeRange(4, sizeof(uint32_t))];
  [result getBytes:&decodedValue range:NSMakeRange(8, sizeof(int32_t))];

  XCTAssertEqual(dictionaryKey, TestEmptyDictionaryKey, @"Dictionary key should match");
  XCTAssertEqual(argumentType, TestInt32ArgumentType, @"Type should be Int32ArgumentType (3)");
  XCTAssertEqual(decodedValue, 42, @"Decoded value should match the original");
}

- (void)testArgumentDataForInt32_WithNegativeValue_PreservesSign
{
  int32_t value = -1;

  NSData *result = [FBInstrumentsClient argumentDataForInt32:value];

  XCTAssertEqual(result.length, (NSUInteger)12);

  int32_t decodedValue = 0;
  [result getBytes:&decodedValue range:NSMakeRange(8, sizeof(int32_t))];
  XCTAssertEqual(decodedValue, -1, @"Negative value should be preserved");
}

- (void)testArgumentDataForInt32_WithZero_ProducesCorrectFormat
{
  NSData *result = [FBInstrumentsClient argumentDataForInt32:0];

  XCTAssertEqual(result.length, (NSUInteger)12);

  int32_t decodedValue = 1;
  [result getBytes:&decodedValue range:NSMakeRange(8, sizeof(int32_t))];
  XCTAssertEqual(decodedValue, 0, @"Zero value should be preserved");
}

- (void)testArgumentDataForInt32_WithMaxValue_PreservesBoundary
{
  int32_t value = INT32_MAX;

  NSData *result = [FBInstrumentsClient argumentDataForInt32:value];

  int32_t decodedValue = 0;
  [result getBytes:&decodedValue range:NSMakeRange(8, sizeof(int32_t))];
  XCTAssertEqual(decodedValue, INT32_MAX, @"INT32_MAX should be preserved");
}

#pragma mark - auxillaryDataFromArgumentsData: Tests

- (void)testAuxillaryDataFromArgumentsData_WithNilArguments_ReturnsEmptyData
{
  NSData *result = [FBInstrumentsClient auxillaryDataFromArgumentsData:nil];

  XCTAssertNotNil(result);
  XCTAssertEqual(result.length, (NSUInteger)0, @"Nil arguments should produce empty data");
}

- (void)testAuxillaryDataFromArgumentsData_WithSingleArgument_IncludesMagicAndLength
{
  NSData *argData = [FBInstrumentsClient argumentDataForArgument:@"test"];

  NSData *result = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[argData]];

  XCTAssertNotNil(result);
  XCTAssertEqual(result.length, (NSUInteger)(16 + argData.length), @"Result should be header (16) + argument data");

  uint64_t magic = 0;
  uint64_t payloadLength = 0;
  [result getBytes:&magic range:NSMakeRange(0, sizeof(uint64_t))];
  [result getBytes:&payloadLength range:NSMakeRange(8, sizeof(uint64_t))];

  XCTAssertEqual(magic, TestArgumentMagic, @"Magic should match ArgumentMagic constant (0x1F0)");
  XCTAssertEqual(payloadLength, (uint64_t)argData.length, @"Payload length should match argument data length");
}

- (void)testAuxillaryDataFromArgumentsData_WithMultipleArguments_ConcatenatesCorrectly
{
  NSData *arg1 = [FBInstrumentsClient argumentDataForArgument:@"first"];
  NSData *arg2 = [FBInstrumentsClient argumentDataForArgument:@"second"];

  NSData *result = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[arg1, arg2]];

  uint64_t payloadLength = 0;
  [result getBytes:&payloadLength range:NSMakeRange(8, sizeof(uint64_t))];
  XCTAssertEqual(payloadLength, (uint64_t)(arg1.length + arg2.length), @"Payload length should be sum of all argument lengths");
  XCTAssertEqual(result.length, (NSUInteger)(16 + arg1.length + arg2.length));
}

- (void)testAuxillaryDataFromArgumentsData_WithEmptyArray_ProducesHeaderOnly
{
  NSData *result = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[]];

  XCTAssertEqual(result.length, (NSUInteger)16, @"Empty array should produce header-only data (16 bytes)");

  uint64_t payloadLength = 0;
  [result getBytes:&payloadLength range:NSMakeRange(8, sizeof(uint64_t))];
  XCTAssertEqual(payloadLength, (uint64_t)0, @"Payload length should be zero for empty array");
}

#pragma mark - objectArgumentsFromAuxillaryData:error: Tests

- (void)testObjectArgumentsFromAuxillaryData_RoundTrip_WithSingleString
{
  NSData *argData = [FBInstrumentsClient argumentDataForArgument:@"hello"];
  NSData *auxData = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[argData]];

  NSError *error = nil;
  NSArray<id> *result = [FBInstrumentsClient objectArgumentsFromAuxillaryData:auxData error:&error];

  XCTAssertNil(error, @"Round-trip should not produce an error");
  XCTAssertNotNil(result);
  XCTAssertEqual(result.count, (NSUInteger)1, @"Should contain exactly one argument");
  XCTAssertEqualObjects(result[0], @"hello", @"Decoded argument should match original");
}

- (void)testObjectArgumentsFromAuxillaryData_RoundTrip_WithMultipleArguments
{
  NSData *arg1 = [FBInstrumentsClient argumentDataForArgument:@"first"];
  NSData *arg2 = [FBInstrumentsClient argumentDataForArgument:@(42)];
  NSData *auxData = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[arg1, arg2]];

  NSError *error = nil;
  NSArray<id> *result = [FBInstrumentsClient objectArgumentsFromAuxillaryData:auxData error:&error];

  XCTAssertNil(error);
  XCTAssertEqual(result.count, (NSUInteger)2);
  XCTAssertEqualObjects(result[0], @"first");
  XCTAssertEqualObjects(result[1], @(42));
}

- (void)testObjectArgumentsFromAuxillaryData_WithInsufficientData_ReturnsError
{
  uint8_t shortBytes[] = {0x00, 0x01, 0x02, 0x03, 0x04, 0x05};
  NSData *shortData = [NSData dataWithBytes:shortBytes length:sizeof(shortBytes)];

  NSError *error = nil;
  NSArray<id> *result = [FBInstrumentsClient objectArgumentsFromAuxillaryData:shortData error:&error];

  XCTAssertNil(result, @"Should return nil for insufficient data");
  XCTAssertNotNil(error, @"Should produce an error for insufficient data");
}

- (void)testObjectArgumentsFromAuxillaryData_RoundTrip_WithDictionary
{
  NSDictionary *dict = @{@"key" : @"value", @"number" : @(123)};
  NSData *argData = [FBInstrumentsClient argumentDataForArgument:dict];
  NSData *auxData = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[argData]];

  NSError *error = nil;
  NSArray<id> *result = [FBInstrumentsClient objectArgumentsFromAuxillaryData:auxData error:&error];

  XCTAssertNil(error);
  XCTAssertEqual(result.count, (NSUInteger)1);
  XCTAssertEqualObjects(result[0], dict, @"Decoded dictionary should match original");
}

- (void)testObjectArgumentsFromAuxillaryData_PreservesArgumentOrder
{
  NSData *arg1 = [FBInstrumentsClient argumentDataForArgument:@"alpha"];
  NSData *arg2 = [FBInstrumentsClient argumentDataForArgument:@"beta"];
  NSData *arg3 = [FBInstrumentsClient argumentDataForArgument:@"gamma"];
  NSData *auxData = [FBInstrumentsClient auxillaryDataFromArgumentsData:@[arg1, arg2, arg3]];

  NSError *error = nil;
  NSArray<id> *result = [FBInstrumentsClient objectArgumentsFromAuxillaryData:auxData error:&error];

  XCTAssertNil(error);
  XCTAssertEqual(result.count, (NSUInteger)3);
  XCTAssertEqualObjects(result[0], @"alpha", @"First argument should be alpha");
  XCTAssertEqualObjects(result[1], @"beta", @"Second argument should be beta");
  XCTAssertEqualObjects(result[2], @"gamma", @"Third argument should be gamma");
}

#pragma mark - advanceData: Tests

- (void)testAdvanceData_WithBuffer_ReadsCorrectBytesAndAdvances
{
  uint32_t values[] = {0xDEADBEEF, 0xCAFEBABE, 0x12345678};
  NSData *data = [NSData dataWithBytes:values length:sizeof(values)];

  uint32_t firstValue = 0;
  NSData *remaining = [FBInstrumentsClient advanceData:data buffer:&firstValue length:sizeof(uint32_t)];

  XCTAssertEqual(firstValue, (uint32_t)0xDEADBEEF, @"First value should be read correctly");
  XCTAssertEqual(remaining.length, (NSUInteger)(sizeof(values) - sizeof(uint32_t)), @"Remaining data should be shorter");

  uint32_t secondValue = 0;
  NSData *remaining2 = [FBInstrumentsClient advanceData:remaining buffer:&secondValue length:sizeof(uint32_t)];
  XCTAssertEqual(secondValue, (uint32_t)0xCAFEBABE);
  XCTAssertEqual(remaining2.length, (NSUInteger)sizeof(uint32_t));
}

- (void)testAdvanceData_WithDataOut_ExtractsSubdataAndAdvances
{
  uint8_t bytes[] = {0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08};
  NSData *data = [NSData dataWithBytes:bytes length:sizeof(bytes)];

  NSData *extracted = nil;
  NSData *remaining = [FBInstrumentsClient advanceData:data dataOut:&extracted length:3];

  XCTAssertNotNil(extracted);
  XCTAssertEqual(extracted.length, (NSUInteger)3);
  uint8_t expectedExtracted[] = {0x01, 0x02, 0x03};
  XCTAssertEqualObjects(extracted, [NSData dataWithBytes:expectedExtracted length:3]);
  XCTAssertEqual(remaining.length, (NSUInteger)5, @"Remaining should have 5 bytes left");
}

- (void)testAdvanceData_WithNilDataOut_StillAdvances
{
  uint8_t bytes[] = {0x01, 0x02, 0x03, 0x04, 0x05};
  NSData *data = [NSData dataWithBytes:bytes length:sizeof(bytes)];

  NSData *remaining = [FBInstrumentsClient advanceData:data dataOut:nil length:2];

  XCTAssertEqual(remaining.length, (NSUInteger)3, @"Should advance past 2 bytes even with nil dataOut");
}

@end

File: FBDeviceControlTests/Tests/Unit/FBInstrumentsClientTests.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import FBControlCore
@testable import FBDeviceControl
import Foundation
import Testing

// Constants mirroring private definitions in FBInstrumentsClient.
private let TestArgumentMagic: UInt64 = 0x1F0
private let TestEmptyDictionaryKey: UInt32 = 10
private let TestObjectArgumentType: UInt32 = 2
private let TestInt32ArgumentType: UInt32 = 3

private func readUInt32(_ data: Data, at offset: Int) -> UInt32 {
  var value: UInt32 = 0
  withUnsafeMutableBytes(of: &value) { destination in
    destination.copyBytes(from: data.subdata(in: offset..<(offset + 4)))
  }
  return value
}

private func readInt32(_ data: Data, at offset: Int) -> Int32 {
  Int32(bitPattern: readUInt32(data, at: offset))
}

private func readUInt64(_ data: Data, at offset: Int) -> UInt64 {
  var value: UInt64 = 0
  withUnsafeMutableBytes(of: &value) { destination in
    destination.copyBytes(from: data.subdata(in: offset..<(offset + 8)))
  }
  return value
}

/// The DTX argument wire format: what goes out for each argument kind, and that the auxillary
/// framing round-trips.
@Suite
struct FBInstrumentsClientTests {

  // MARK: - Object arguments

  @Test
  func argumentDataForStringProducesValidSerializedData() throws {
    let argument = "com.example.testapp"

    let result = FBInstrumentsClient.argumentData(forArgument: argument)

    #expect(result.count > 12, "Result should be larger than the 12-byte header")
    #expect(readUInt32(result, at: 0) == TestEmptyDictionaryKey)
    #expect(readUInt32(result, at: 4) == TestObjectArgumentType)
    let argumentSize = readUInt32(result, at: 8)
    #expect(Int(argumentSize) == result.count - 12, "Argument size should match remaining data length")

    let archived = result.subdata(in: 12..<(12 + Int(argumentSize)))
    let decoded = try NSKeyedUnarchiver.unarchivedObject(ofClass: NSString.self, from: archived)
    #expect(decoded as String? == argument)
  }

  @Test
  func argumentDataForDictionaryProducesValidSerializedData() throws {
    let argument: [String: Bool] = ["StartSuspendedKey": true, "KillExisting": false]

    let result = FBInstrumentsClient.argumentData(forArgument: argument)

    #expect(result.count > 12)
    let argumentSize = readUInt32(result, at: 8)
    let archived = result.subdata(in: 12..<(12 + Int(argumentSize)))
    let decoded = try NSKeyedUnarchiver.unarchivedObject(
      ofClasses: [NSDictionary.self, NSNumber.self, NSString.self], from: archived)
    #expect(decoded as? [String: Bool] == argument)
  }

  @Test
  func argumentDataForEmptyStringProducesValidSerializedData() throws {
    let result = FBInstrumentsClient.argumentData(forArgument: "")

    #expect(result.count > 12)
    let argumentSize = readUInt32(result, at: 8)
    let archived = result.subdata(in: 12..<(12 + Int(argumentSize)))
    let decoded = try NSKeyedUnarchiver.unarchivedObject(ofClass: NSString.self, from: archived)
    #expect(decoded as String? == "")
  }

  // MARK: - Int32 arguments

  @Test
  func argumentDataForInt32WithPositiveValueProducesCorrectFormat() {
    let result = FBInstrumentsClient.argumentData(forInt32: 42)

    #expect(result.count == 12, "Int32 argument data should be exactly 12 bytes")
    #expect(readUInt32(result, at: 0) == TestEmptyDictionaryKey)
    #expect(readUInt32(result, at: 4) == TestInt32ArgumentType)
    #expect(readInt32(result, at: 8) == 42)
  }

  @Test
  func argumentDataForInt32WithNegativeValuePreservesSign() {
    let result = FBInstrumentsClient.argumentData(forInt32: -1)

    #expect(result.count == 12)
    #expect(readInt32(result, at: 8) == -1)
  }

  @Test
  func argumentDataForInt32WithZeroProducesCorrectFormat() {
    let result = FBInstrumentsClient.argumentData(forInt32: 0)

    #expect(result.count == 12)
    #expect(readInt32(result, at: 8) == 0)
  }

  @Test
  func argumentDataForInt32WithMaxValuePreservesBoundary() {
    let result = FBInstrumentsClient.argumentData(forInt32: Int32.max)

    #expect(readInt32(result, at: 8) == Int32.max)
  }

  // MARK: - Auxillary framing

  @Test
  func auxillaryDataFromNilArgumentsReturnsEmptyData() {
    let result = FBInstrumentsClient.auxillaryData(fromArgumentsData: nil)

    #expect(result.isEmpty, "Nil arguments should produce empty data")
  }

  @Test
  func auxillaryDataFromSingleArgumentIncludesMagicAndLength() {
    let argData = FBInstrumentsClient.argumentData(forArgument: "test")

    let result = FBInstrumentsClient.auxillaryData(fromArgumentsData: [argData])

    #expect(result.count == 16 + argData.count, "Result should be header (16) + argument data")
    #expect(readUInt64(result, at: 0) == TestArgumentMagic)
    #expect(readUInt64(result, at: 8) == UInt64(argData.count))
  }

  @Test
  func auxillaryDataFromMultipleArgumentsConcatenatesCorrectly() {
    let arg1 = FBInstrumentsClient.argumentData(forArgument: "first")
    let arg2 = FBInstrumentsClient.argumentData(forArgument: "second")

    let result = FBInstrumentsClient.auxillaryData(fromArgumentsData: [arg1, arg2])

    #expect(readUInt64(result, at: 8) == UInt64(arg1.count + arg2.count))
    #expect(result.count == 16 + arg1.count + arg2.count)
  }

  @Test
  func auxillaryDataFromEmptyArrayProducesHeaderOnly() {
    let result = FBInstrumentsClient.auxillaryData(fromArgumentsData: [])

    #expect(result.count == 16, "Empty array should produce header-only data (16 bytes)")
    #expect(readUInt64(result, at: 8) == 0)
  }

  // MARK: - Round trips

  @Test
  func objectArgumentsRoundTripWithSingleString() throws {
    let argData = FBInstrumentsClient.argumentData(forArgument: "hello")
    let auxData = FBInstrumentsClient.auxillaryData(fromArgumentsData: [argData])

    let result = try FBInstrumentsClient.objectArguments(fromAuxillaryData: auxData)

    #expect(result.count == 1)
    #expect(result.first as? String == "hello")
  }

  @Test
  func objectArgumentsRoundTripWithMultipleArguments() throws {
    let arg1 = FBInstrumentsClient.argumentData(forArgument: "first")
    let arg2 = FBInstrumentsClient.argumentData(forArgument: 42)
    let auxData = FBInstrumentsClient.auxillaryData(fromArgumentsData: [arg1, arg2])

    let result = try FBInstrumentsClient.objectArguments(fromAuxillaryData: auxData)

    #expect(result.count == 2)
    #expect(result.first as? String == "first")
    #expect(result.last as? Int == 42)
  }

  @Test
  func objectArgumentsWithInsufficientDataThrows() {
    let shortData = Data([0x00, 0x01, 0x02, 0x03, 0x04, 0x05])

    #expect(throws: (any Error).self) {
      try FBInstrumentsClient.objectArguments(fromAuxillaryData: shortData)
    }
  }

  @Test
  func objectArgumentsRoundTripWithDictionary() throws {
    let dictionary: [String: Any] = ["key": "value", "number": 123]
    let argData = FBInstrumentsClient.argumentData(forArgument: dictionary)
    let auxData = FBInstrumentsClient.auxillaryData(fromArgumentsData: [argData])

    let result = try FBInstrumentsClient.objectArguments(fromAuxillaryData: auxData)

    #expect(result.count == 1)
    let decoded = result.first as? NSDictionary
    #expect(decoded == dictionary as NSDictionary)
  }

  // MARK: - The advanceData compatibility walkers

  @Test
  func advanceDataWithBufferReadsCorrectBytesAndAdvances() {
    let values: [UInt32] = [0xDEAD_BEEF, 0xCAFE_BABE, 0x1234_5678]
    let data = values.withUnsafeBytes { Data($0) }

    var firstValue: UInt32 = 0
    let remaining = withUnsafeMutableBytes(of: &firstValue) { buffer in
      FBInstrumentsClient.advance(data, buffer: buffer.baseAddress!, length: 4)
    }

    #expect(firstValue == 0xDEAD_BEEF)
    #expect(remaining.count == 8)

    var secondValue: UInt32 = 0
    let remaining2 = withUnsafeMutableBytes(of: &secondValue) { buffer in
      FBInstrumentsClient.advance(remaining, buffer: buffer.baseAddress!, length: 4)
    }
    #expect(secondValue == 0xCAFE_BABE)
    #expect(remaining2.count == 4)
  }

  @Test
  func advanceDataWithDataOutExtractsSubdataAndAdvances() {
    let data = Data([0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08])

    var extracted: NSData?
    let remaining = FBInstrumentsClient.advance(data, dataOut: &extracted, length: 3)

    #expect(extracted as Data? == Data([0x01, 0x02, 0x03]))
    #expect(remaining.count == 5)
  }

  @Test
  func advanceDataWithNilDataOutStillAdvances() {
    let data = Data([0x01, 0x02, 0x03, 0x04, 0x05])

    let remaining = FBInstrumentsClient.advance(data, dataOut: nil, length: 2)

    #expect(remaining.count == 3, "Should advance past 2 bytes even with nil dataOut")
  }

  @Test
  func objectArgumentsPreservesArgumentOrder() throws {
    let auxData = FBInstrumentsClient.auxillaryData(fromArgumentsData: [
      FBInstrumentsClient.argumentData(forArgument: "alpha"),
      FBInstrumentsClient.argumentData(forArgument: "beta"),
      FBInstrumentsClient.argumentData(forArgument: "gamma"),
    ])

    let result = try FBInstrumentsClient.objectArguments(fromAuxillaryData: auxData)

    #expect(result.count == 3)
    #expect(result.first as? String == "alpha")
    #expect(result.dropFirst().first as? String == "beta")
    #expect(result.last as? String == "gamma")
  }
}
